server/admin.go

```diff
@@ -668,21 +668,31 @@ func (s *adminServer) Drain(ctx context.Context, req *serverpb.DrainRequest) (*s
 
 	nowOn, err := s.server.Drain(on)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	nowOnInts := make([]int32, len(nowOn))
+	res := serverpb.DrainResponse{
+		On: make([]int32, len(nowOn)),
+	}
 	for i := range nowOn {
-		nowOnInts[i] = int32(nowOn[i])
+		res.On[i] = int32(nowOn[i])
+	}
+	if err := stream.Send(&res); err != nil {
+		return err
 	}
+
 	if req.Shutdown {
-		s.server.stopper.Quiesce()
-		go func() {
-			time.Sleep(50 * time.Millisecond)
-			s.server.stopper.Stop()
-		}()
+		go s.server.stopper.Stop()
+	}
+
+	ctx := stream.Context()
+
+	select {
+	case <-s.server.stopper.IsStopped():
```

Comment:
If the stopper closes the connection first and then stops some other stuff after, why does the client know that the server has completely terminated by the time its connection is closed? I see what the intention of the code is, but I don't see at all why it works.